ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java

```diff
@@ -760,6 +760,41 @@ void seal() throws StoreException {
     persistBloomFilter();
   }
 
+  /**
+   * @return index value of last PUT record in this index segment. Return {@code null} if no PUT is found
+   */
+  IndexValue getIndexValueOfLastPut() throws StoreException {
+    IndexValue indexValueOfLastPut = null;
+    if (sealed.get()) {
+      ByteBuffer readBuf = serEntries.duplicate();
+      int numOfIndexEntries = numberOfEntries(readBuf);
+      NavigableSet<IndexValue> values = new TreeSet<>();
+      for (int i = 0; i < numOfIndexEntries; i++) {
+        StoreKey key = getKeyAt(readBuf, i);
+        values.clear();
+        getAllValuesFromMmap(readBuf, key, i, numOfIndexEntries, values);
+        for (IndexValue indexValue : values) {
+          // FLAGS_DEFAULT_VALUE means PUT record
+          if (indexValue.getFlags() == IndexValue.FLAGS_DEFAULT_VALUE && (indexValueOfLastPut == null
+              || indexValue.compareTo(indexValueOfLastPut) > 0)) {
+            indexValueOfLastPut = indexValue;
```

Comment:
Partially agree. True, I could add a `break` in the loop of  `for (IndexValue indexValue : values)` when PUT record is found.
Note that, the navigable set `values` contains all index entries associated with specific `key`. So, maximum size of this set is 3 ( one PUT entry, one TTL Update entry and one DELETE entry). Due to the nature of log semantic, PUT always comes first. Hence, there is no need to search in reverse order.

Reply:
So having a loop is just a safety mechanic (since you could technically just grab the first entry and throw an exception if it's not a PUT)?

Reply:
right, I probably could check first entry only but the current way doesn't introduce any more overhead. Technically, we can throw exception if it's not a PUT, which means the store is in a "incorrect" state violating log/index semantics. However, 1. this situation shouldn't occur with Ambry current implementation. 2. throwing exception makes transition fail because this method is called when decommissioning a replica. It's probably ok and replica will be put into ERROR state but I feel like not necessary because the intention here is to ensure peer replicas catch up with local replica, the order of PUT entries could be fixed by other peer replicas and local one is still safe to be removed. (My two cents)